ray-curator/ray_curator/stages/modules/filter.py

```diff
@@ -0,0 +1,500 @@
+
+    def setup_on_node(self, _node_info: NodeInfo, _worker_metadata: WorkerMetadata) -> None:
+        if self.name in ["lang_id", "fasttext_quality_filter"]:
+            if not os.path.exists(self.score_fn._model_path):
+                msg = f"Model file {self.score_fn._model_path} not found"
+                raise FileNotFoundError(msg)
+        elif self.name in ["token_count"]:
+            # Use snapshot_download to download all files without loading the model into memory.
+            huggingface_hub.snapshot_download(
+                repo_id=self.score_fn._hf_model_name,
+                token=self.score_fn._hf_token,
+                local_files_only=False,  # Download if not cached
+                resume_download=True,  # Resume interrupted downloads
+            )
+
+    def setup(self, _: WorkerMetadata) -> None:
+        if self.name in ["lang_id", "fasttext_quality_filter"]:
+            self.score_fn.model = fasttext.load_model(self.score_fn._model_path)
+        elif self.name in ["token_count"] and self.score_fn._tokenizer is None:
+            self.score_fn._tokenizer = AutoTokenizer.from_pretrained(
+                self.score_fn._hf_model_name,
+                local_files_only=True,  # Fail if not cached
+            )
+
+    def process_batch(self, tasks: list[DocumentBatch]) -> list[DocumentBatch]:
+        """
+        Scores all records in multiple dataset batches
+
+        Args:
+            tasks (list[DocumentBatch]): The dataset batches to apply the module to
+
+        Returns:
+            list[DocumentBatch]: The dataset batches with the score applied
+
+        """
+
+        if self.processing_batch_size == 1:
+            return [self.process(tasks[0])]
+
+        # Convert and collect all Pandas DataFrames
+        dfs = [task.to_pandas() for task in tasks]
+        lengths = [len(df) for df in dfs]
+        task_ids = [task.task_id for task in tasks]
+        dataset_names = [task.dataset_name for task in tasks]
+
+        # Combine and process
+        combined_df = pd.concat(dfs, ignore_index=True)
+        combined_batch = DocumentBatch(data=combined_df, task_id="batch_list", dataset_name="batch_list")
+        processed_batch = self.process(combined_batch)
+        processed_df = processed_batch.to_pandas()
+
+        # Use original lengths to rebuild the DocumentBatch objects
+        result_batches = []
+        offset = 0
+        for length, task_id, dataset_name in zip(lengths, task_ids, dataset_names, strict=True):
+            chunk = processed_df.iloc[offset : offset + length].reset_index(drop=True)
+            result_batches.append(
+                DocumentBatch(
+                    task_id=f"{task_id}_{self.name}",
+                    dataset_name=dataset_name,
+                    data=chunk,
+                )
+            )
+            offset += length
+
+        return result_batches
+
+    def process(self, batch: DocumentBatch) -> DocumentBatch | None:
```

Comment:
If you go ahead with process_batch then I believe you shouldn't have to override `process(..)` since process is only ever called by process_batch, in which case I'd recommend removing this and straight up calling at line 115

Reply:
I think it could. I am thinking about 2 cases:

1. A lot of `DocumentBatch` objects, each with relatively small DataFrames, would benefit with `process_batch`.
2. A few `DocumentBatch` objects, each with large DataFrames, would benefit with just `process`.